app/Domain/Goalcanvas/Services/Goalcanvas.php

```diff
@@ -202,5 +223,488 @@ public function pollForUpdatedGoals(): array|false
+            $progressValue = $item['currentValue'] - $item['startValue'];
+
+            return $total > 0 ? round($progressValue / $total * 100, 2) : 0;
+        }
+
+        private function determineCurrentCanvasId(array $allCanvas, array $params): int
+        {
+            $sessionKey = "current" . strtoupper(static::CANVAS_NAME) . "Canvas";
+
+            if (isset($params['id'])) {
+                $currentCanvasId = (int)$params['id'];
+            } elseif (isset($_REQUEST['searchCanvas'])) {
+                $currentCanvasId = (int)$_REQUEST['searchCanvas'];
+            } elseif (session()->exists($sessionKey)) {
+                $currentCanvasId = session($sessionKey);
+                $currentCanvasId = $this->validateCanvasId($currentCanvasId, $allCanvas);
+            } else {
+                $currentCanvasId = $allCanvas[0]['id'] ?? -1;
+            }
+
+            session([$sessionKey => $currentCanvasId]);
+
+            return $currentCanvasId;
+        }
+
+        private function validateCanvasId(int $canvasId, array $allCanvas): int
+        {
+            foreach ($allCanvas as $canvas) {
+                if ($canvasId == $canvas['id']) {
+                    return $canvasId;
+                }
+            }
+            return -1;
+        }
+
+        // Goal Dashboard Post
+
+        public function handleDashboardPostRequest($params): array
+        {
+            $result = [
+                'redirect' => false,
+                'redirectUrl' => '',
+                'notification' => null,
+            ];
+
+            $currentCanvasId = session("current" . strtoupper(static::CANVAS_NAME) . "Canvas") ?? -1;
+
+            if (isset($params['newCanvas'])) {
+                $result = $this->handleNewCanvas($params, $result);
+            } elseif (isset($params['editCanvas']) && $currentCanvasId > 0) {
+                $result = $this->handleEditCanvas($params, $currentCanvasId, $result);
+            } elseif (isset($params['cloneCanvas']) && $currentCanvasId > 0) {
+                $result = $this->handleCloneCanvas($params, $currentCanvasId, $result);
+            } elseif (isset($params['mergeCanvas']) && $currentCanvasId > 0) {
+                $result = $this->handleMergeCanvas($params, $currentCanvasId, $result);
+            } elseif (isset($params['importCanvas'])) {
+                $result = $this->handleImportCanvas($params, $result);
+            }
+
+            return $result;
+        }
+
+        private function handleNewCanvas($params, $result): array
+        {
+            if (!isset($params['canvastitle']) || empty($params['canvastitle'])) {
+                return $this->setNotification($result, 'notification.please_enter_title', 'error');
+            }
+
+            if ($this->goalRepository->existCanvas(session("currentProject"), $params['canvastitle'])) {
+                return $this->setNotification($result, 'notification.board_exists', 'error');
```

Comment:
Please move template notifications into the controller. The service should just return success or failure messages. (Maybe with failure errors/exceptions) 
So in this case you could throw a "ElementExistsException()" (just create a new exception type into core/exceptions. The controller can then decide what to do with that exception. (aka show a notification)